src/Twine/Guard.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module Twine.Guard (
+    TerminationAction (..)
+  , TerminationHandler (..)
+  , guarded
+  , repeatedly
+  ) where
+
+import           Control.Monad.Catch
+import           Control.Monad.Trans.Either
+import           Control.Monad.Trans.Class
+
+import           P
+
+import           System.IO
+
+import           Twine.Snooze
+
+data TerminationAction =
+    Restart
+  | Die
+  deriving (Eq, Show)
+
+data TerminationHandler e =
+  TerminationHandler {
+      onExplosion :: SomeException -> IO TerminationAction
+    , onError :: e -> IO TerminationAction
+    , onGraceful :: IO TerminationAction
+    }
+
+-- | Run an action forever, using termination handler to notify of
```

Comment:
I still find the wording a bit ambiguous because the "forever" part of running depends on the `onGraceful` function. If `onGraceful` returns `Die` then we stop.

Wouldn't that be better to propose:

 - `guarded` with no `onGraceful`, just running the action to completion or not but offering the possibility of a restart
 - `repeatedly` with no `onGraceful` as well but recursing after a delay (and allowing restart as well)